shared-libs/cht-datasource/src/contact-types.ts

```diff
@@ -0,0 +1,78 @@
+import { DataObject, Identifiable, isDataObject, isIdentifiable, Nullable } from './libs/core';
+import { Doc } from './libs/doc';
+import {
+  byContactType,
+  byFreetext,
+  ContactTypeQualifier,
+  FreetextQualifier,
+  isUuidQualifier,
+  UuidQualifier
+} from './qualifier';
+import { InvalidArgumentError } from './libs/error';
+
+/** @ignore */
+export namespace v1 {
+  /** @internal */
+  export interface NormalizedParent extends DataObject, Identifiable {
+    readonly parent?: NormalizedParent;
+  }
+
+  /** @internal */
+  export interface Contact extends Doc, NormalizedParent {
+    readonly contact_type?: string;
+    readonly name?: string;
+    readonly reported_date?: Date;
+    readonly type: string;
+  }
+
+  /** @internal */
+  export interface ContactWithLineage extends Contact {
+    readonly parent?: ContactWithLineage | NormalizedParent;
+  }
+  
+  /** @ignore */
+  export const isNormalizedParent = (value: unknown): value is NormalizedParent => {
+    return isDataObject(value) && isIdentifiable(value) && (!value.parent || isNormalizedParent(value.parent));
+  };
+
+  /** @ignore */
+  export const isContactType = (value: ContactTypeQualifier | FreetextQualifier): value is ContactTypeQualifier => {
```

Comment:
Not exactly a duplicate but a subset I'd say. 
We have been using the following pattern in our code to validate the arguments into `src/<entity>.ts`.
<img src="https://github.com/user-attachments/assets/873e6eda-e39c-4179-abb8-86efeeccd395" width="300">
I've removed the use of the local `isContactType` function for the screenshot in which case both `type` and `freetext` are turned into required params. We could write another `assert*` function that does the combination of `type` or `freetext` like `assertTypeOrFreetextQualifier` but for some reason, the current implementation seemed more appropriate because the control is in the caller's hands. I do not have a strong inclination towards the current implementation though. Both seem equally logical and valid. I still agree that functions like these would be more appropriate in a single file but this function seemed small enough to be kept here.